addons/mail/models/res_users.py

```diff
@@ -117,43 +117,54 @@ def _unsubscribe_from_channels(self):
 
     @api.model
     def systray_get_activities(self):
-        activities = self.env["mail.activity"].search([("user_id", "=", self.env.uid)])
-        activities_by_record_by_model_name = defaultdict(lambda: defaultdict(lambda: self.env["mail.activity"]))
-        for activity in activities:
-            record = self.env[activity.res_model].browse(activity.res_id)
-            activities_by_record_by_model_name[activity.res_model][record] += activity
-        model_ids = list({self.env["ir.model"]._get(name).id for name in activities_by_record_by_model_name.keys()})
+        query = """SELECT array_agg(res_id) as res_ids, m.id, count(*), act.res_model as model,
+                    CASE
+                        WHEN %(today)s::date - act.date_deadline::date = 0 Then 'today'
+                        WHEN %(today)s::date - act.date_deadline::date > 0 Then 'overdue'
+                        WHEN %(today)s::date - act.date_deadline::date < 0 Then 'planned'
+                    END AS states
+                FROM mail_activity AS act
+                JOIN ir_model AS m ON act.res_model_id = m.id
+                WHERE user_id = %(user_id)s
+                GROUP BY m.id, states, act.res_model;
+                """
+        self.env.cr.execute(query, {
+            'today': fields.Date.context_today(self),
+            'user_id': self.env.uid,
+        })
+        activity_data = self.env.cr.dictfetchall()
+        dic = defaultdict(lambda: {"today": set(), "overdue": set(), "planned": set(), "all": set()})
+        for data in activity_data:
+            dic[data["id"]][data["states"]] = set(data["res_ids"])
+            dic[data["id"]]["all"] = dic[data["id"]]["all"] | set(data["res_ids"])
         user_activities = {}
-        for model_name, activities_by_record in activities_by_record_by_model_name.items():
-            domain = [("id", "in", list({r.id for r in activities_by_record.keys()}))]
-            allowed_records = self.env[model_name].search(domain)
+        for model_id in dic:
+            model_dic = dic[model_id]
+            model = self.env["ir.model"].browse(model_id).with_prefetch(tuple(dic.keys()))
+            allowed_records = self.env[model.model].search([("id", "in", tuple(model_dic["all"]))])
             if not allowed_records:
                 continue
-            module = self.env[model_name]._original_module
+            allowed_records = allowed_records.mapped("id")
```

Comment:
I would not rewrite the variable with a different kind of value here, just using `allowed_records.ids` in code should do it.